### PR TITLE
Let devs intercept onChanged events and control when to update their adapter

### PR DESCRIPTION
**Pull request by SUPERCILEX**
Would address issues like #482 and see https://github.com/firebase/FirebaseUI-Android/issues/15#issuecomment-265644228.

### CODE REVIEW

File: app/src/main/java/com/firebase/uidemo/database/ChatActivity.java [hunk 1]
```diff
@@ -93,11 +94,7 @@ public void onComplete(DatabaseError databaseError, DatabaseReference reference)
         });
 
         mMessages = (RecyclerView) findViewById(R.id.messagesList);
-
         mManager = new LinearLayoutManager(this);
-        mManager.setReverseLayout(false);
-
-        mMessages.setHasFixedSize(false);
```
Comment (SUPERCILEX) [inline app/src/main/java/com/firebase/uidemo/database/ChatActivity.java]: Unnecessary, these are defaults.

File: app/src/main/java/com/firebase/uidemo/database/ChatActivity.java [hunk 2]
```diff
@@ -169,17 +165,25 @@ public void onItemRangeInserted(int positionStart, int itemCount) {
...
+                })
+                .addOnCompleteListener(new OnCompleteListener<AuthResult>() {
                     @Override
                     public void onComplete(@NonNull Task<AuthResult> task) {
-                        Log.d(TAG, "signInAnonymously:onComplete:" + task.isSuccessful());
                         if (task.isSuccessful()) {
-                            Toast.makeText(ChatActivity.this, "Signed In",
-                                    Toast.LENGTH_SHORT).show();
-                            attachRecyclerViewAdapter();
+                            Toast.makeText(getApplicationContext(),
```
Comment (SUPERCILEX) [inline app/src/main/java/com/firebase/uidemo/database/ChatActivity.java]: Show toast and continue logging in even is screen was rotated.

**Comment by SUPERCILEX** [comment]
@puf PTAL, these changes shouldn't be anything major and can be merged in this branch since I'm targeting the dev branch. I'll merge your other PRs into this branch and resolve conflicts since I think there will be style issues causing the Travis build to fail. (FirebaseUI now has static analysis checks!)

File: app/src/main/java/com/firebase/uidemo/database/ChatActivity.java (hunk 2 above)
Comment (samtstern) [inline app/src/main/java/com/firebase/uidemo/database/ChatActivity.java]: Why did we remove the call to `attachRecyclerViewAdapter()` here?  Also if we scope the onComplete listener to the Activity context then we can avoid dead activity issues.

**Comment by samtstern** [comment]
@puf could you take a look?  Basically this does two things:

1) Move the `OnChangedListener` interface out of `FirebaseArray` and make it first class.
2) Make a `protected onChanged()` function in the adapters so that some developers can change the behavior.

File: app/src/main/java/com/firebase/uidemo/database/ChatActivity.java (hunk 2 above)
Comment (SUPERCILEX) [inline app/src/main/java/com/firebase/uidemo/database/ChatActivity.java]: @samtstern I've extracted the toast logic into a static class to prevent activity leaks. As for the `attachRecyclerViewAdapter`, I moved that up to an onSuccess listener that is tied to the activity.

**Comment by SUPERCILEX** [comment]
@puf Since `OnChangedListener` is now public would you be ok with me renaming things?

1. `OnChangedListener` -> **`ChangeEventListener`** to be more consistent with Firebase apis
2. `onChanged` -> **`onChildChanged`** as discussed in https://github.com/firebase/FirebaseUI-Android/pull/487#discussion_r94525452

@puf @samtstern What do you guys think?

File: app/src/main/java/com/firebase/uidemo/database/ChatActivity.java (hunk 1 above)
Comment (puf) [inline app/src/main/java/com/firebase/uidemo/database/ChatActivity.java]: There are valid reasons for having this code here, which we could debate here. But these reasons have nothing to do with letting devs intercept change events. In future PRs please keep functional changes separate from code cleanup.

File: app/src/main/java/com/firebase/uidemo/database/ChatActivity.java [hunk 3]
```diff
@@ -108,10 +106,10 @@ public void onStart() {
         // Default Database rules do not allow unauthenticated reads, so we need to
         // sign in before attaching the RecyclerView adapter otherwise the Adapter will
         // not be able to read any data from the Database.
-        if (!isSignedIn()) {
-            signInAnonymously();
-        } else {
+        if (isSignedIn()) {
             attachRecyclerViewAdapter();
+        } else {
```
Comment (puf) [inline app/src/main/java/com/firebase/uidemo/database/ChatActivity.java]: Same as before: this change seems unrelated to the topic of this PR and just expresses a personal preference. While that preference might well be better than the existing code, it should not pollute this PR.

File: app/src/main/java/com/firebase/uidemo/database/ChatActivity.java [hunk 4]
```diff
@@ -286,4 +276,21 @@ public void setText(String text) {
             mTextField.setText(text);
         }
     }
+
+    private static class SignInResultNotifier implements OnCompleteListener<AuthResult> {
```
Comment (puf) [inline app/src/main/java/com/firebase/uidemo/database/ChatActivity.java]: Nice addition! I'd appreciate some documentation on this class though.

File: database/src/main/java/com/firebase/ui/database/OnChangedListener.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
+package com.firebase.ui.database;
+
+import com.google.firebase.database.ChildEventListener;
+import com.google.firebase.database.DataSnapshot;
+import com.google.firebase.database.DatabaseError;
+
+public interface OnChangedListener {
```
Comment (puf) [inline database/src/main/java/com/firebase/ui/database/OnChangedListener.java]: I'm fine pulling this up. But wouldn't it be better as an interface in FirebaseArray (which I see moving towards becoming an abstract base-class or interface itself)? Having this inside FirebaseArray makes it more explicitly scoped as "these operations apply to this FirebaseArray (or subclass) object".

Comment (SUPERCILEX) [inline database/src/main/java/com/firebase/ui/database/OnChangedListener.java]: @puf, I would like that too, but you've chosen to keep `FirebaseArray` package private so people will get an error when overriding `onChanged` saying they can't access `FirebaseArray.OnChangedListener.EventType`. To put `OnChangedListener` in `FirebaseArray` we would have to go public and I don't think `FirebaseArray` is quite ready for that.

File: app/src/main/java/com/firebase/uidemo/database/ChatActivity.java (hunk 1 above)
Comment (SUPERCILEX) [inline app/src/main/java/com/firebase/uidemo/database/ChatActivity.java]: I removed this change, but I am curious, why are we setting default values and adding seemingly unnecessary code?

File: app/src/main/java/com/firebase/uidemo/database/ChatActivity.java (hunk 3 above)
Comment (SUPERCILEX) [inline app/src/main/java/com/firebase/uidemo/database/ChatActivity.java]: @puf I write my PRs with the philosophy that small improvements as we go along are better than constantly submitting side PRs for this kind of change. However, I see where you are coming from and removed these changes from this PR.

**Comment by puf** [comment]
@SUPERCILEX: Not sure why we had these layout properties in there, but I
know for a fact I referred developers specifically to this part of the
example multiple times. It's the same as explicitly initializing a variable
to its default: you make it clear that it's not mistake and can more easily
refer to it to discuss it.

**Comment by SUPERCILEX** [comment]
I see, it makes sense now. @puf Thanks for letting me know!

**Comment by SUPERCILEX** [comment]
@puf since you pushed 21d5d44397a670d617b7ff2e41996d733ff122f1, I went ahead with the [rename](https://github.com/firebase/FirebaseUI-Android/pull/488#issuecomment-270735029).

**Comment by samtstern** [comment]
I am going to start doing QA on version `1.1.0` as there are many important fixes in there and it's been longer than expected.  I have targeted this PR for `1.2.0` which will be the next minor release.

**Comment by SUPERCILEX** [comment]
@puf, any chance you could take another look at this? (I just merged everything)

File: database/src/main/java/com/firebase/ui/database/OnChangedListener.java (hunk 5 above)
Comment (puf) [inline database/src/main/java/com/firebase/ui/database/OnChangedListener.java]: But then isn't this essentially doing the same: making a part of the internal workings of FirebaseArray public?

File: database/src/main/java/com/firebase/ui/database/FirebaseArray.java [hunk 6]
```diff
@@ -80,21 +69,21 @@ public void onChildAdded(DataSnapshot snapshot, String previousChildKey) {
             index = getIndexForKey(previousChildKey) + 1;
         }
         mSnapshots.add(index, snapshot);
-        notifyChangedListeners(OnChangedListener.EventType.ADDED, index);
+        notifyChangedListeners(ChangeEventListener.EventType.ADDED, index);
```
Comment (puf) [inline database/src/main/java/com/firebase/ui/database/FirebaseArray.java]: Definitely a better name! +1

File: database/src/main/java/com/firebase/ui/database/FirebaseListAdapter.java [hunk 7]
```diff
@@ -51,7 +51,7 @@
  *            contained in the children of the given Firebase location
  */
 public abstract class FirebaseListAdapter<T> extends BaseAdapter {
-    private static final String TAG = FirebaseListAdapter.class.getSimpleName();
+    private static final String TAG = "FirebaseListAdapter";
```
Comment (puf) [inline database/src/main/java/com/firebase/ui/database/FirebaseListAdapter.java]: Can you explain why this change is needed/useful?

**Comment by puf** [comment]
@SUPERCILEX I left a few comments. Most are very minor. The only real concern is that we'll end up exposing internal details of the `FirebaseArray` when those aren't sufficiently hardened yet. But I'm very willing to take a vote there, as the recent changes feel like a move in the right direction.

Of course we'll need the same ability to listen for changes in FirebaseUI-iOS, before we can merge. @morganchen12 did you already have a look at this?

File: database/src/main/java/com/firebase/ui/database/FirebaseListAdapter.java (hunk 7 above)
Comment (SUPERCILEX) [inline database/src/main/java/com/firebase/ui/database/FirebaseListAdapter.java]: @puf This isn't so much needed in `FirebaseListAdapter`, but this is a massive no-no in the [`FirebaseRecyclerAdapter`](https://github.com/firebase/FirebaseUI-Android/pull/488/files/17a136d524ad4840f49bea54ba740016b7245b19#diff-724a30131c3284a3a9d723d6bb38561dR76) where the name is 23 characters long. I made this tag in one of my first PRs and I really regret it: if someone were to create a tag using `getSimpleName()` in `FirebaseIndexRecyclerAdapter`, `Log.anything` will [throw an exception](http://stackoverflow.com/questions/28168622/the-logging-tag-can-be-at-most-23-characters) because the tag name exceeds 24 characters. You should _**NEVER**_ use dynamic tag names for that reason.

@puf, since this is higher priority than #488, I've create #525.

**Comment by SUPERCILEX** [comment]
@puf SGTM! I think you'll like my upcoming PRs for `FirebaseArray`... 😄

**Comment by morganchen12** [comment]
I think this feature is added with the [transforms PR](https://github.com/firebase/FirebaseUI-iOS/pull/186) on iOS.

**Comment by SUPERCILEX** [comment]
@puf @morganchen12 @samtstern This PR doesn't provide explicit support for joins because @puf [doesn't want that](https://github.com/firebase/FirebaseUI-Android/issues/15#issuecomment-264465745), but it does allow devs to do their own filtering as seen in [my crappy example](https://github.com/firebase/FirebaseUI-Android/issues/15#issuecomment-265644228). So, @morganchen12 @puf do you guys want to merge this PR and https://github.com/firebase/FirebaseUI-iOS/pull/186 and call it good because @puf doesn't want to support on device filerting? Or should we do something else?

**Comment by morganchen12** [comment]
iOS doesn't add on-device filtering either, so no feature parity issues there.

**Comment by SUPERCILEX** [comment]
@samtstern @puf @morganchen12 So are we good to merge?

**Comment by samtstern** [comment]
Still fine with me, will let @puf pull the trigger.  I have just pushed a new `version-1.2.0-dev` branch off of `master` where we should put this feature.

**Comment by morganchen12** [comment]
Fine with me, pending @puf's decision as well

**Comment by SUPERCILEX** [comment]
@puf, could you please take another look at this? It's blocking progress on my other PRs... Thanks!

**Comment by samtstern** [comment]
Ok after some team discussion it's time to merge this.  Squashing and merging into the `1.2.0` branch.

**Comment by SUPERCILEX** [comment]
@samtstern Really glad to have this merged!

I started writing a post demoing some use cases so just in case anyone wants to see it, here you go: https://github.com/firebase/FirebaseUI-Android/issues/482#issuecomment-274956357.
